### PR TITLE
Rff sampler

### PR DESCRIPTION
**Pull request by henrymoss**
I have added a method to the GP prior so that we can get approximate samples from the GP.  I use the standard RFF approach to get samples with functional forms that can be cheaply queried as many times as we like.

### CODE REVIEW

File: gpjax/gps.py [hunk 1]
```diff
@@ -101,6 +103,13 @@ def _initialise_params(self, key: KeyArray) -> Dict:
         return self.init_params(key)
 
 
+FunctionalSample = Callable[[Float[Array, "N D"]], Float[Array, "N B"]]
+""" Type alias for functions representing `B` samples from a model, to be evaluated on any set of
+`N` inputs (of dimension `D`) and returning the evaluations of each (potentially approximate)
+sample draw across these inputs.
+"""
+
+
```
Comment (thomaspinder) [inline gpjax/gps.py:112]: Nice. Could this go in a `utils` or `types` file?

File: gpjax/gps.py [hunk 2]
```diff
@@ -247,6 +254,89 @@ def predict_fn(
...
         return predict_fn
 
+    def sample_approx(
+        self,
+        num_samples: int,
+        params: Dict,
+        seed: KeyArray,
+        num_features: Optional[int] = 100,
+    ) -> FunctionalSample:
+        """Build an approximate sample from the Gaussian process prior. This method
```
Comment (thomaspinder) [inline gpjax/gps.py:264]: Think you'll need a raw string or double backslashes.

File: gpjax/gps.py [hunk 3]
```diff
@@ -247,6 +254,89 @@ def predict_fn(
...
+            kernel.
+
+
+        Returns:
+            FunctionalSample: A function representing an approximate sample from the Gaussian
+            process prior.
+        """
+        for integer_input in [num_features, num_samples]:
+            if (not isinstance(integer_input, int)) or integer_input < 0:
+                raise ValueError
```
Comment (thomaspinder) [inline gpjax/gps.py:314]: Can you be explicit here e.g., `ValueError('input is negative and must be positive')` or `ValueError(f'input is of type {type(integer_input} and must be an integer')`?

**Review by thomaspinder — CHANGES_REQUESTED** [review]
Good PR. Documentation and testing is excellent. Few minor comments to address and then we can merge.

File: gpjax/gps.py (hunk 2 above)
Comment (henrymoss) [reply gpjax/gps.py:274]: where?

Comment (thomaspinder) [reply gpjax/gps.py:258]: `"""Build` should be `r"""Build`

File: gpjax/gps.py [hunk 4]
```diff
@@ -466,12 +552,124 @@ def predict(test_inputs: Float[Array, "N D"]) -> GaussianDistribution:
...
 
+    def sample_approx(
+        self,
+        num_samples: int,
+        params: Dict,
+        train_data: Dataset,
+        key: KeyArray,
+        num_features: Optional[int] = 100,
+    ) -> FunctionalSample:
+        """Build an approximate sample from the Gaussian process posterior. This method
```
Comment (thomaspinder) [inline gpjax/gps.py:567]: Same comment as above re. `r"""`

File: gpjax/gps.py [hunk 5]
```diff
@@ -466,12 +552,124 @@ def predict(test_inputs: Float[Array, "N D"]) -> GaussianDistribution:
...
 
         return predict
 
+    def sample_approx(
+        self,
+        num_samples: int,
+        params: Dict,
+        train_data: Dataset,
+        key: KeyArray,
+        num_features: Optional[int] = 100,
```
Comment (thomaspinder) [inline gpjax/gps.py:565]: Is 100 sensible here?

**Review by thomaspinder — APPROVED** [review]

**Pull request merged**